File: 84-LargestRectangleInHistogram/84-LargestRectangleInHistogram.py
# Last updated: 8/13/2025, 8:23:02 PM
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:

        nser = [len(heights)]*len(heights)
        nsel = [-1]*len(heights)

        self.findNSER(heights , nser)
        self.findNSEL(heights , nsel)

        max_rectangle = float('-inf')

        for i in range(len(heights)) : 
            smaller_on_right = nser[i] - 1
            smaller_on_left = nsel[i] + 1

            max_rectangle = max(max_rectangle , heights[i] * (smaller_on_right - smaller_on_left + 1))
        
        return max_rectangle
    

    def findNSER(self , heights : list[int] , nser : list[int]) -> None : 

        stack = []
        for i in range(len(heights)) : 
            while len(stack) > 0 and stack[-1][0] > heights[i] : 
                popped = stack.pop(-1)
                index = popped[1]

                nser[index] = i

            stack.append([heights[i] , i])
        
        print(nser)
        
    def findNSEL(self , heights : list[int] , nsel : list[int]) -> None : 
        stack = []
        for i in range(len(heights)-1 , -1 , -1) : 
            while len(stack) > 0 and stack[-1][0] > heights[i] : 
                popped = stack.pop(-1)
                index = popped[1]

                nsel[index] = i

            stack.append([heights[i] , i])
        
        print(nsel)

File: 84-LargestRectangleInHistogram/84-LargestRectangleInHistogram.java
// Last updated: 8/13/2025, 8:23:07 PM
class Solution {
    public int largestRectangleArea(int[] heights) {
        Stack<Integer> s = new Stack();
        int ans = Integer.MIN_VALUE;
        int nsel[] = new int[heights.length];
        int nser[] = new int[heights.length];
        int n = heights.length;

        for(int i=0;i<n;i++){
            while(s.size()>0 && heights[i]<heights[s.peek()]){
                int temp = s.pop();
                nser[temp] = i;
            }
            s.add(i);
        }
        while(s.size()>0){
            int temp = s.pop();
            nser[temp] = heights.length;
        }

        for(int i=n-1;i>=0;i--){
            while(s.size()>0 && heights[i]<heights[s.peek()]){
                int temp = s.pop();
                nsel[temp] = i;
            }
            s.add(i);
        }

        while(s.size()>0){
            int temp = s.pop();
            nsel[temp] = -1;
        }

        for(int i=0;i<n;i++){
            int val1 = nser[i] - 1;
            int val2 = nsel[i] + 1;
            int wd = val1 - val2 +1;
            int h = heights[i];
            int area = wd*h;
            ans = Math.max(area,ans);
        }

        return ans;
    }
}